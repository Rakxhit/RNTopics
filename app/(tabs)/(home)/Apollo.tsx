import { Button, SafeAreaView, StyleSheet, Text, View } from "react-native";
import React from "react";
import { gql, useQuery, useSuspenseQuery } from "@apollo/client";
const POLLING_SECONDS = 5000;

const Apollo = () => {
  const GET_USER_DATE = gql`
    query {
      company {
        ceo
      }
    }
  `;

  const { loading, error, data, startPolling, stopPolling } = useQuery(
    GET_USER_DATE,
    {
      pollInterval: POLLING_SECONDS, // This api will get called every single second
      notifyOnNetworkStatusChange: true, // This will regenerate the data instead of showing the same data on the refetch and the loading state will also be maintained
      variables: {
        name: "users",
      },
    }
  );

  const { data: suspenseData } = useSuspenseQuery(GET_USER_DATE);
  console.log("suspenseData: ", suspenseData);

  if (error) return <Text>What an error</Text>;
  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.mainContainer}>
        <Text>Default polling is {POLLING_SECONDS}</Text>
        <Text style={{ opacity: loading ? 0.2 : 1, paddingVertical: 20 }}>
          {data?.company?.ceo}
        </Text>
        <Button
          testID="start-polling"
          title="Start polling on 1 second"
          onPress={() => {
            startPolling(1000);
          }}
        />
        <Button
          testID="stop-polling"
          title="Stop polling "
          onPress={() => {
            stopPolling();
          }}
        />
      </View>
    </SafeAreaView>
  );
};

export default Apollo;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    backgroundColor: "white",
  },
  mainContainer: {
    alignItems: "center",
    paddingTop: 30,
  },
});
