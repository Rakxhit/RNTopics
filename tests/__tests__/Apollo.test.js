import Apollo from "@/app/(tabs)/(home)/Apollo";
import { fireEvent, render, screen } from "@testing-library/react-native";
import { useQuery } from "@apollo/client";

jest.mock("@apollo/client", () => {
  const originalModule = jest.requireActual("@apollo/client");
  return {
    __esModule: true,
    ...originalModule,
    useSuspenseQuery: jest.fn().mockReturnValue({
      data: {
        company: {
          ceo: "raxit",
        },
      },
    }),
    useQuery: jest.fn().mockReturnValue({
      loading: false,
      error: "",
      data: {
        company: {
          ceo: "raxit",
        },
      },
      startPolling: () => jest.fn(),
      stopPolling: () => jest.fn(),
    }),
  };
});

describe("Apollo component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render", () => {
    const { toJSON } = render(<Apollo />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("should have the Desired text", () => {
    const title = "Default polling is 5000";
    const { getByText } = render(<Apollo />);
    expect(getByText(title)).toBeTruthy();
  });

  it("it should return the error message", () => {
    jest.mocked(useQuery).mockReturnValue({
      loading: false,
      error: new Error("Some Error Occured"),
      data: null,
      startPolling: jest.fn(),
      stopPolling: jest.fn(),
    });

    const { getByText } = render(<Apollo />);
    expect(getByText("What an error")).toBeTruthy();
  });

  it("calling start polling with 1000", () => {
    const startPolling = jest.fn();
    const stopPolling = jest.fn();
    jest.mocked(useQuery).mockReturnValue({
      loading: false,
      error: "",
      data: {
        company: {
          ceo: "raxit",
        },
      },
      startPolling: startPolling,
      stopPolling: stopPolling,
    });

    const { getByTestId } = render(<Apollo />);

    const startPollingButton = getByTestId("start-polling");
    expect(startPollingButton).toBeTruthy();
    fireEvent.press(startPollingButton);
    expect(startPolling).toHaveBeenCalled();

    const stopPollingButton = getByTestId("stop-polling");
    expect(stopPollingButton).toBeTruthy();
    fireEvent.press(stopPollingButton);
    expect(stopPolling).toHaveBeenCalled();
  });
});
