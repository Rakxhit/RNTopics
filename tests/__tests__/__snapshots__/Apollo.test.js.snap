// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apollo component should render 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "paddingTop": 30,
      }
    }
  >
    <Text>
      Default polling is 
      5000
    </Text>
    <Text
      style={
        {
          "opacity": 1,
          "paddingVertical": 20,
        }
      }
    >
      raxit
    </Text>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="start-polling"
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Start polling on 1 second
        </Text>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="stop-polling"
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Stop polling 
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
